File: Dataset/postagger/pSCRDRtagger/RDRPOSTagger.py
# -*- coding: utf-8 -*-

import os,sys,time

#Set Python & directory paths
os.chdir("../")
sys.setrecursionlimit(100000)  
sys.path.append(os.path.abspath(""))
os.chdir("./pSCRDRtagger")
    
from SCRDRlearner.PosTaggingRDRTree import PosTaggingRDRTree
from Utility.Utils import getWordTag, getRawTextFromFile, readDictionary
from SCRDRlearner.Object import FWObject
from InitialTagger.InitialTagger import InitTagger4Corpus, InitTagger4Sentence

#Change or add other thresholds: thresholds = [(5,5),(3, 2),(2,2)]
thresholds = [(3, 2)]

class SampleRDRTree(PosTaggingRDRTree):
    """
    RDRPOSTagger for a particular language
    """
    def __init__(self):
        self.root = None
    
    def tagRawSentence(self, DICT, rawLine):
        line = InitTagger4Sentence(DICT, rawLine)
        sen = ''
        wordTags = line.replace("“", "''").replace("”", "''").replace("\"", "''").split()
        for i in xrange(len(wordTags)):
            fwObject = FWObject.getFWObject(wordTags, i)
            word, tag = getWordTag(wordTags[i])
            node = self.findFiredNode(fwObject)
            if node.depth > 0:
                sen += word + "/" + node.conclusion + " "
            else: #Fired at root, return initialized tag
                sen += word + "/" + tag + " "
        return sen.strip()

    def tagRawCorpus(self, DICT, rawCorpusInputPath):
        print '\nTagging raw word-segmented corpus: %s' % rawCorpusInputPath
        outW = open(rawCorpusInputPath + ".TAGGED", "w")
        for line in open(rawCorpusInputPath, "r").readlines():
            outW.write(self.tagRawSentence(DICT, line) + "\n")  
        outW.close()
        print "Done!"

def SampleTraining(args = sys.argv[2:]):
    pathToDict = args[0]
    dirPath = os.path.join(args[1] + "/")
    correctTrain = args[2]
    learntRules = args[3]
    
    print '\nTraining RDRPOSTagger for a particular language-driven POS Tagging...'   
    print "Initial tagging..."
    
    #Get raw text from golden corpus
    getRawTextFromFile(dirPath + correctTrain, dirPath + correctTrain + ".RAW")
    DICT = readDictionary(pathToDict)
    #Calling initial tagger for a specific language:
    InitTagger4Corpus(DICT, dirPath + correctTrain + ".RAW", dirPath + correctTrain + ".INIT")
    
    print "Done Initialization!"
    
    print 'Building SCRDR-based POS tagging tree of rules...'
    
    for (improveThreshold, matchThreshold) in thresholds:
        timeStart = time.time()
        outputDir = "T%d-%d/" % (improveThreshold, matchThreshold)
        os.mkdir(dirPath + outputDir)         
        
        rdrTree = PosTaggingRDRTree(improveThreshold, matchThreshold) 
        rdrTree.buildTreeFromCorpus(dirPath + correctTrain + ".INIT", dirPath + correctTrain)
        
        print "Write the tree to file..."
        rdrTree.writeToFileWithoutSeenCases(dirPath + outputDir + learntRules)
        #rdrTree.writeToFile(dirPath + outputDir + learntRules)
        
        print "\nTraining time for threshold %d-%d: %f seconds\n" % (improveThreshold, matchThreshold, time.time() - timeStart)
            
    print '\nCompleted!'

def printInstructions():
    print '\nRDRPOSTagger instructions for a specific language:'
    print '\nTo train RDRPOSTagger for POS tagging:'
    print '\npython RDRPOSTagger.py train PATH-TO-LEXICON PATH-TO-DIRECTORY GOLDEN-CORPUS MODEL-NAME'
    print '\nExample: python RDRPOSTagger.py train ../Sample/En/shortDict ../Sample/En correctTrain postagging.rdr'
    print '\nTo tag a raw word-segmented corpus using the trained tagger:'
    print '\npython RDRPOSTagger.py tag PATH-TO-TRAINED-MODEL PATH-TO-LEXICON PATH-TO-RAW-CORPUS'
    print '\nExample: python RDRPOSTagger.py tag ../Sample/En/T3-2/postagging.rdr ../Sample/En/fullDict ../Sample/En/rawTest'
    print '\nHaving a look at http://rdrpostagger.sourceforge.net to find more information!!!'
    
def runSampleRDRPOSTagger(args = sys.argv[1:]):
    if (len(args) == 0):
        printInstructions()
    elif args[0].lower().find("train") > -1:
        SampleTraining()
    elif args[0].lower().find("tag") > -1:
        r = SampleRDRTree()
        r.constructTreeFromRulesFile(args[1])
        DICT = readDictionary(args[2])
        r.tagRawCorpus(DICT, args[3])
    else:
        printInstructions()
        
if __name__ == "__main__":
    runSampleRDRPOSTagger()
    pass

File: Dataset/postagger/jSCRDRtagger/RDRPOSTagger.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.HashMap;
import java.util.List;

/**
 * @author DatQuocNguyen
 * 
 */
public class RDRPOSTagger
{
	public Node root;

	public RDRPOSTagger()
	{
	}

	public RDRPOSTagger(Node node)
	{
		root = node;
	}

	// Build an scrdr-based tree for pos tagging from a learned model file
	// containing rules
	public void constructTreeFromRulesFile(String rulesFilePath)
		throws IOException
	{
		BufferedReader buffer = new BufferedReader(new InputStreamReader(
				new FileInputStream(new File(rulesFilePath)), "UTF-8"));
		String line = buffer.readLine();

		this.root = new Node(new FWObject(false), "NN", null, null, null, 0);

		Node currentNode = this.root;
		int currentDepth = 0;

		for (; (line = buffer.readLine()) != null;) {
			int depth = 0;
			for (int i = 0; i <= 5; i++) { // supposed that the maximum
											// exception level is up to 5.
				if (line.charAt(i) == '\t')
					depth += 1;
				else
					break;
			}

			line = line.trim();
			if (line.length() == 0)
				continue;

			if (line.contains("cc:"))
				continue;

			// System.out.println(line);
			FWObject condition = Utils
					.getCondition(line.split(" : ")[0].trim());
			String conclusion = Utils
					.getConclusion(line.split(" : ")[1].trim());

			Node node = new Node(condition, conclusion, null, null, null, depth);

			if (depth > currentDepth) {
				currentNode.setExceptNode(node);
			}
			else if (depth == currentDepth) {
				currentNode.setIfnotNode(node);
			}
			else {
				while (currentNode.depth != depth)
					currentNode = currentNode.fatherNode;
				currentNode.setIfnotNode(node);
			}
			node.setFatherNode(currentNode);

			currentNode = node;
			currentDepth = depth;
		}
		buffer.close();
	}

	public Node findFiredNode(FWObject object)
	{
		Node currentN = root;
		Node firedN = null;
		while (true) {
			if (currentN.satisfy(object)) {
				firedN = currentN;
				if (currentN.exceptNode == null) {
					break;
				}
				else {
					currentN = currentN.exceptNode;
				}
			}
			else {
				if (currentN.ifnotNode == null) {
					break;
				}
				else {
					currentN = currentN.ifnotNode;
				}
			}

		}

		return firedN;
	}

	public String tagInitializedSentence(String inInitializedSentence)
	{
		StringBuilder sb = new StringBuilder();
		// List of pair (word, tag) in the input initialized sentence
		List<WordTag> wordtags = Utils.getWordTagList(inInitializedSentence);
		int size = wordtags.size();
		for (int i = 0; i < size; i++) {
			FWObject object = Utils.getObject(wordtags, size, i);
			Node firedNode = findFiredNode(object);
			if (firedNode.depth > 0)
				sb.append(wordtags.get(i).word + "/" + firedNode.conclusion
						+ " ");
			else {
				// Fired at root, return initialized tag.
				sb.append(wordtags.get(i).word + "/" + wordtags.get(i).tag
						+ " ");
			}
		}
		return sb.toString();
	}

	// Tag an initialized corpus. The initialized corpus are already generated
	// (before) by using an external initial tagger.
	public void tagInitializedCorpus(String inInitializedFilePath)
		throws IOException
	{
		BufferedReader buffer = new BufferedReader(new InputStreamReader(
				new FileInputStream(new File(inInitializedFilePath)), "UTF-8"));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(
				new FileOutputStream(inInitializedFilePath + ".TAGGED"),
				"UTF-8"));

		for (String line; (line = buffer.readLine()) != null;) {
			line = line.trim();
			if (line.length() == 0) {
				bw.write("\n");
				continue;
			}
			bw.write(tagInitializedSentence(line) + "\n");
		}
		buffer.close();
		bw.close();
	}

	public String tagVnSentence(HashMap<String, String> FREQDICT,
			String sentence)
		throws IOException
	{
		StringBuilder sb = new StringBuilder();

		String line = sentence.trim();
		if (line.length() == 0) {
			return "\n";
		}

		line = line.replace("“", "''").replace("”", "''").replace("\"", "''");
		// Call initial tagger for Vietnamese
		List<WordTag> wordtags = InitialTagger.VnInitTagger4Sentence(FREQDICT,
				line);

		int size = wordtags.size();
		for (int i = 0; i < size; i++) {
			FWObject object = Utils.getObject(wordtags, size, i);
			Node firedNode = findFiredNode(object);
			sb.append(wordtags.get(i).word + "/" + firedNode.conclusion + " ");
		}

		return sb.toString();
	}

	public void tagVnCorpus(HashMap<String, String> FREQDICT,
			String inRawFilePath)
		throws IOException
	{
		BufferedReader buffer = new BufferedReader(new InputStreamReader(
				new FileInputStream(new File(inRawFilePath)), "UTF-8"));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(
				new FileOutputStream(inRawFilePath + ".TAGGED"), "UTF-8"));
		for (String line; (line = buffer.readLine()) != null;) {
			bw.write(tagVnSentence(FREQDICT, line) + "\n");
		}
		buffer.close();
		bw.close();
	}

	public String tagEnSentence(HashMap<String, String> FREQDICT,
			String sentence)
		throws IOException
	{
		StringBuilder sb = new StringBuilder();

		String line = sentence.trim();
		if (line.length() == 0) {
			return "\n";
		}

		line = line.replace("“", "''").replace("”", "''").replace("\"", "''");
		// Call initial tagger for English
		List<WordTag> wordtags = InitialTagger.EnInitTagger4Sentence(FREQDICT,
				line);

		int size = wordtags.size();
		for (int i = 0; i < size; i++) {
			FWObject object = Utils.getObject(wordtags, size, i);
			Node firedNode = findFiredNode(object);
			sb.append(wordtags.get(i).word + "/" + firedNode.conclusion + " ");
		}
		return sb.toString();
	}

	public void tagEnCorpus(HashMap<String, String> FREQDICT,
			String inRawFilePath)
		throws IOException
	{
		BufferedReader buffer = new BufferedReader(new InputStreamReader(
				new FileInputStream(new File(inRawFilePath)), "UTF-8"));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(
				new FileOutputStream(inRawFilePath + ".TAGGED"), "UTF-8"));
		for (String line; (line = buffer.readLine()) != null;) {
			bw.write(tagEnSentence(FREQDICT, line) + "\n");
		}
		buffer.close();
		bw.close();
	}

	public String tagSentence(HashMap<String, String> FREQDICT, String sentence)
		throws IOException
	{
		StringBuilder sb = new StringBuilder();
		String line = sentence.trim();
		if (line.length() == 0) {
			return "\n";
		}

		line = line.replace("“", "''").replace("”", "''").replace("\"", "''");
		// Calling the initial tagger for a particular language
		List<WordTag> wordtags = InitialTagger.InitTagger4Sentence(FREQDICT,
				line);
		int size = wordtags.size();
		for (int i = 0; i < size; i++) {
			FWObject object = Utils.getObject(wordtags, size, i);
			Node firedNode = findFiredNode(object);
			if (firedNode.depth > 0)
				sb.append(wordtags.get(i).word + "/" + firedNode.conclusion
						+ " ");
			else {
				// Fired at root, return initialized tag.
				sb.append(wordtags.get(i).word + "/" + wordtags.get(i).tag
						+ " ");
			}
		}
		return sb.toString();
	}

	public void tagCorpus(HashMap<String, String> FREQDICT, String inRawFilePath)
		throws IOException
	{
		BufferedReader buffer = new BufferedReader(new InputStreamReader(
				new FileInputStream(new File(inRawFilePath)), "UTF-8"));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(
				new FileOutputStream(inRawFilePath + ".TAGGED"), "UTF-8"));
		for (String line; (line = buffer.readLine()) != null;) {
			bw.write(tagSentence(FREQDICT, line) + "\n");
		}
		buffer.close();
		bw.close();

	}

	public static void printInstructions()
	{
		System.out.println("Run RDRPOSTagger for tagging a corpus:");
		System.out
				.println("java RDRPOSTagger OPTION PATH-TO-LEARNED-MODEL [PATH-TO-LEXICON] PATH-TO-CORPUS");
		System.out
				.println("in which OPTION gets one of the 4 values 'en', 'vn', 'other', 'init' "
						+ "corresponding to a tagging process on English, Vietnamese, other languages"
						+ "and on an initialized corpus, respectively.");
		System.out
				.println("Example1: java RDRPOSTagger other ../Sample/En/T3-2/postagging.rdr ../Sample/En/fullDict ../Sample/En/rawTest");
		System.out
				.println("Example2: java RDRPOSTagger en ../Models/EN.RDR ../Dicts/EN.DICT ../Sample/En/rawTest");
		System.out
				.println("Example3: java RDRPOSTagger vn ../Models/VN.RDR ../Dicts/VN.DICT ../Sample/Vn/rawTest");
		System.out
				.println("Example4: java RDRPOSTagger init ../Sample/En/T3-2/postagging.rdr ../Sample/En/initTest");
		System.out
				.println("Having a look at http://rdrpostagger.sourceforge.net to find more information!");

	}

	public static void runRDRPOSTagging(String args[])
		throws IOException
	{
		if (args.length <= 1)
			printInstructions();
		else {
			RDRPOSTagger tree = new RDRPOSTagger();
			System.out
					.println("\nBuilding SCRDR-based tree for POS tagging from file: "
							+ args[1]);
			tree.constructTreeFromRulesFile(args[1]);

			if (args[0].equals("en")) {
				System.out.println("Reading lexicon: " + args[2]);
				HashMap<String, String> FREQDICT = Utils.getDictionary(args[2]);
				System.out
						.println("Tagging raw word-segmented corpus in English:"
								+ args[3]);
				tree.tagEnCorpus(FREQDICT, args[3]);
			}
			else if (args[0].equals("vn")) {
				System.out.println("Reading lexicon: " + args[2]);
				HashMap<String, String> FREQDICT = Utils.getDictionary(args[2]);
				System.out
						.println("Tagging raw word-segmented corpus in Vietnamese:"
								+ args[3]);
				tree.tagVnCorpus(FREQDICT, args[3]);
			}
			else if (args[0].equals("other")) {
				System.out.println("Reading lexicon: " + args[2]);
				HashMap<String, String> FREQDICT = Utils.getDictionary(args[2]);
				System.out.println("Tagging raw word-segmented corpus:"
						+ args[3]);
				tree.tagCorpus(FREQDICT, args[3]);
			}
			else if (args[0].equals("init")) {
				System.out.println("Tagging initialized corpus:" + args[2]);
				tree.tagInitializedCorpus(args[2]);
			}
			else {
				printInstructions();
			}
		}
	}

	public static void main(String[] args)
		throws IOException
	{
		runRDRPOSTagging(args);
	}
}
